File: frontend/scripts/quality-check.js
#!/usr/bin/env node
/**
 * Frontend Code Quality Check Script
 * Runs all code quality checks for the frontend codebase.
 */

const { execSync } = require('child_process');
const path = require('path');
const fs = require('fs');

// Color codes for terminal output
const colors = {
  green: '\x1b[92m',
  red: '\x1b[91m',
  yellow: '\x1b[93m',
  blue: '\x1b[94m',
  reset: '\x1b[0m',
};

function printHeader(message) {
  console.log(`\n${colors.blue}${'='.repeat(80)}${colors.reset}`);
  console.log(`${colors.blue}${message.padStart((80 + message.length) / 2).padEnd(80)}${colors.reset}`);
  console.log(`${colors.blue}${'='.repeat(80)}${colors.reset}\n`);
}

function printSuccess(message) {
  console.log(`${colors.green}✓ ${message}${colors.reset}`);
}

function printError(message) {
  console.log(`${colors.red}✗ ${message}${colors.reset}`);
}

function printWarning(message) {
  console.log(`${colors.yellow}⚠ ${message}${colors.reset}`);
}

function runCommand(command, description, allowFailure = false) {
  console.log(`\n${colors.yellow}Running: ${description}${colors.reset}`);
  console.log(`Command: ${command}`);

  try {
    execSync(command, {
      stdio: 'inherit',
      cwd: path.join(__dirname, '..'),
    });
    printSuccess(`${description} passed`);
    return true;
  } catch (error) {
    if (allowFailure) {
      printWarning(`${description} failed (non-blocking)`);
      return false;
    } else {
      printError(`${description} failed`);
      return false;
    }
  }
}

function checkESLint() {
  return runCommand(
    'npm run lint',
    'ESLint code linting'
  );
}

function checkPrettier() {
  return runCommand(
    'npm run format:check',
    'Prettier formatting check'
  );
}

function checkTypeScript() {
  return runCommand(
    'npm run type-check',
    'TypeScript type checking'
  );
}

function checkDependencies() {
  return runCommand(
    'npx depcheck',
    'Unused dependencies check',
    true
  );
}

function checkNpmAudit() {
  return runCommand(
    'npm audit --audit-level=moderate',
    'npm security audit',
    true
  );
}

function checkBundleSize() {
  console.log(`\n${colors.yellow}Running: Bundle size analysis${colors.reset}`);
  
  try {
    // Check if .next directory exists
    const nextDir = path.join(__dirname, '..', '.next');
    if (!fs.existsSync(nextDir)) {
      printWarning('Build directory not found. Run npm run build first.');
      return true; // Don't fail if build doesn't exist
    }

    execSync('npm run build', {
      stdio: 'inherit',
      cwd: path.join(__dirname, '..'),
    });

    printSuccess('Bundle size analysis completed');
    return true;
  } catch (error) {
    printWarning('Bundle size analysis failed (non-blocking)');
    return false;
  }
}

function fixFormatting() {
  printHeader('Auto-fixing Code Formatting');

  const eslintSuccess = runCommand(
    'npm run lint:fix',
    'ESLint auto-fix'
  );

  const prettierSuccess = runCommand(
    'npm run format',
    'Prettier auto-format'
  );

  return eslintSuccess && prettierSuccess;
}

function main() {
  // Parse command line arguments
  const fixMode = process.argv.includes('--fix');

  if (fixMode) {
    printHeader('Frontend Code Quality - Fix Mode');
    if (fixFormatting()) {
      printSuccess('\nFormatting fixes applied successfully!');
      process.exit(0);
    } else {
      printError('\nSome formatting fixes failed');
      process.exit(1);
    }
  }

  printHeader('Frontend Code Quality Checks');

  const results = {};

  // Run all checks
  printHeader('Code Linting');
  results.eslint = checkESLint();

  printHeader('Code Formatting');
  results.prettier = checkPrettier();

  printHeader('Type Checking');
  results.typescript = checkTypeScript();

  printHeader('Dependency Analysis');
  results.dependencies = checkDependencies();

  printHeader('Security Checks');
  results.audit = checkNpmAudit();

  printHeader('Bundle Analysis');
  results.bundle = checkBundleSize();

  // Print summary
  printHeader('Quality Check Summary');

  const passed = Object.values(results).filter(v => v).length;
  const total = Object.keys(results).length;

  for (const [check, success] of Object.entries(results)) {
    if (success) {
      printSuccess(`${check.charAt(0).toUpperCase() + check.slice(1)}: PASSED`);
    } else {
      printError(`${check.charAt(0).toUpperCase() + check.slice(1)}: FAILED`);
    }
  }

  console.log(`\n${colors.blue}Results: ${passed}/${total} checks passed${colors.reset}`);

  if (passed === total) {
    printSuccess('\n✓ All quality checks passed!');
    process.exit(0);
  } else {
    printError(`\n✗ ${total - passed} quality check(s) failed`);
    printWarning('\nRun with --fix to auto-fix formatting issues');
    process.exit(1);
  }
}

main();

File: backend/scripts/quality_check.py
#!/usr/bin/env python3
"""
Backend Code Quality Check Script
Runs all code quality checks for the backend codebase.
"""

import subprocess
import sys
from pathlib import Path
from typing import List, Tuple

# Set UTF-8 encoding for Windows console
if sys.platform == "win32":
    import codecs
    sys.stdout = codecs.getwriter("utf-8")(sys.stdout.detach())
    sys.stderr = codecs.getwriter("utf-8")(sys.stderr.detach())

# Color codes for terminal output
GREEN = "\033[92m"
RED = "\033[91m"
YELLOW = "\033[93m"
BLUE = "\033[94m"
RESET = "\033[0m"

# Symbols (fallback for Windows)
CHECK_MARK = "✓" if sys.platform != "win32" else "+"
CROSS_MARK = "✗" if sys.platform != "win32" else "x"
WARNING_MARK = "⚠" if sys.platform != "win32" else "!"


def print_header(message: str) -> None:
    """Print a formatted header."""
    print(f"\n{BLUE}{'=' * 80}{RESET}")
    print(f"{BLUE}{message.center(80)}{RESET}")
    print(f"{BLUE}{'=' * 80}{RESET}\n")


def print_success(message: str) -> None:
    """Print a success message."""
    print(f"{GREEN}{CHECK_MARK} {message}{RESET}")


def print_error(message: str) -> None:
    """Print an error message."""
    print(f"{RED}{CROSS_MARK} {message}{RESET}")


def print_warning(message: str) -> None:
    """Print a warning message."""
    print(f"{YELLOW}{WARNING_MARK} {message}{RESET}")


def run_command(
    command: List[str], description: str, allow_failure: bool = False
) -> Tuple[bool, str]:
    """
    Run a shell command and return success status.
    
    Args:
        command: Command to run as list of strings
        description: Description of what the command does
        allow_failure: If True, don't fail the entire script on error
        
    Returns:
        Tuple of (success, output)
    """
    print(f"\n{YELLOW}Running: {description}{RESET}")
    print(f"Command: {' '.join(command)}")
    
    try:
        result = subprocess.run(
            command,
            capture_output=True,
            text=True,
            check=False,
            encoding='utf-8',
            errors='replace',
        )
        
        if result.returncode == 0:
            print_success(f"{description} passed")
            return True, result.stdout
        else:
            if allow_failure:
                print_warning(f"{description} failed (non-blocking)")
                print(result.stdout)
                print(result.stderr)
                return False, result.stderr
            else:
                print_error(f"{description} failed")
                print(result.stdout)
                print(result.stderr)
                return False, result.stderr
    except Exception as e:
        print_error(f"Error running {description}: {str(e)}")
        return False, str(e)


def check_black() -> bool:
    """Check code formatting with Black."""
    return run_command(
        ["black", "--check", "--diff", "app/", "--config", "pyproject.toml"],
        "Black formatting check",
    )[0]


def check_isort() -> bool:
    """Check import sorting with isort."""
    return run_command(
        ["isort", "--check-only", "--diff", "app/", "--settings-path", "pyproject.toml"],
        "isort import sorting check",
    )[0]


def check_flake8() -> bool:
    """Run flake8 linting."""
    return run_command(
        ["flake8", "app/", "--count", "--statistics", "--config", ".flake8"],
        "Flake8 linting",
    )[0]


def check_mypy() -> bool:
    """Run mypy type checking."""
    return run_command(
        ["mypy", "app/", "--config-file", "pyproject.toml"],
        "MyPy type checking",
        allow_failure=True,
    )[0]


def check_pylint() -> bool:
    """Run pylint code analysis."""
    return run_command(
        ["pylint", "app/", "--exit-zero"],
        "Pylint code analysis",
        allow_failure=True,
    )[0]


def check_bandit() -> bool:
    """Run bandit security checks."""
    return run_command(
        ["bandit", "-r", "app/", "-f", "txt"],
        "Bandit security check",
        allow_failure=True,
    )[0]


def check_safety() -> bool:
    """Check dependencies for security vulnerabilities."""
    return run_command(
        ["safety", "check", "-r", "requirements.txt"],
        "Safety dependency check",
        allow_failure=True,
    )[0]


def check_complexity() -> bool:
    """Check code complexity with radon."""
    success = True
    
    # Cyclomatic complexity
    result = run_command(
        ["radon", "cc", "app/", "-a", "-nc"],
        "Radon cyclomatic complexity",
        allow_failure=True,
    )
    
    # Maintainability index
    run_command(
        ["radon", "mi", "app/", "-nc"],
        "Radon maintainability index",
        allow_failure=True,
    )
    
    return result[0]


def fix_formatting() -> bool:
    """Auto-fix formatting issues."""
    print_header("Auto-fixing Code Formatting")
    
    black_success = run_command(
        ["black", "app/", "--config", "pyproject.toml"],
        "Black auto-formatting",
    )[0]
    
    isort_success = run_command(
        ["isort", "app/", "--settings-path", "pyproject.toml"],
        "isort auto-sorting",
    )[0]
    
    return black_success and isort_success


def main() -> int:
    """Main entry point."""
    # Change to backend directory
    backend_dir = Path(__file__).parent.parent
    import os
    os.chdir(backend_dir)
    
    # Parse command line arguments
    fix_mode = "--fix" in sys.argv
    
    if fix_mode:
        print_header("Backend Code Quality - Fix Mode")
        if fix_formatting():
            print_success("\nFormatting fixes applied successfully!")
            return 0
        else:
            print_error("\nSome formatting fixes failed")
            return 1
    
    print_header("Backend Code Quality Checks")
    
    results = {}
    
    # Run all checks
    print_header("Code Formatting")
    results["black"] = check_black()
    results["isort"] = check_isort()
    
    print_header("Code Linting")
    results["flake8"] = check_flake8()
    
    print_header("Type Checking")
    results["mypy"] = check_mypy()
    
    print_header("Code Analysis")
    results["pylint"] = check_pylint()
    
    print_header("Security Checks")
    results["bandit"] = check_bandit()
    results["safety"] = check_safety()
    
    print_header("Complexity Analysis")
    results["complexity"] = check_complexity()
    
    # Print summary
    print_header("Quality Check Summary")
    
    passed = sum(1 for v in results.values() if v)
    total = len(results)
    
    for check, success in results.items():
        if success:
            print_success(f"{check.capitalize()}: PASSED")
        else:
            print_error(f"{check.capitalize()}: FAILED")
    
    print(f"\n{BLUE}Results: {passed}/{total} checks passed{RESET}")
    
    if passed == total:
        print_success(f"\n{CHECK_MARK} All quality checks passed!")
        return 0
    else:
        print_error(f"\n{CROSS_MARK} {total - passed} quality check(s) failed")
        print_warning("\nRun with --fix to auto-fix formatting issues")
        return 1


if __name__ == "__main__":
    sys.exit(main())
